File: iosApp/Models/BlogPost.swift
import Foundation

struct BlogPost: Codable, Identifiable, Hashable {
    let id: String
    let date: String
    let title: String
    let imageUrl: String?
    let imageAltText: String?
    let imageSize: ImageSize?
    let subheading: String?
    let link: String?
    let linkText: String?
    let bodyText: String
    let isPublished: Bool
}

struct BlogPostDraft: Codable, Equatable {
    var date: String
    var title: String
    var imageUrl: String?
    var imageAltText: String?
    var imageSize: ImageSize?
    var subheading: String?
    var link: String?
    var linkText: String?
    var bodyText: String

    init(post: BlogPost? = nil) {
        date = post?.date ?? Self.today
        title = post?.title ?? ""
        imageUrl = post?.imageUrl
        imageAltText = post?.imageAltText
        imageSize = post?.imageSize
        subheading = post?.subheading?.plainTextFromHTML
        link = post?.link
        linkText = post?.linkText
        bodyText = post?.bodyText.plainTextFromHTML ?? ""
    }

    private static var today: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: .now)
    }
}

private extension String {
    var plainTextFromHTML: String {
        guard contains("<") else { return self }
        guard let data = data(using: .utf8),
              let attributed = try? NSAttributedString(
                data: data,
                options: [
                    .documentType: NSAttributedString.DocumentType.html,
                    .characterEncoding: String.Encoding.utf8.rawValue,
                ],
                documentAttributes: nil
              ) else {
            return replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
        }
        return attributed.string.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

enum ImageSize: String, Codable, CaseIterable, Identifiable {
    case small = "SMALL"
    case medium = "MEDIUM"
    case large = "LARGE"
    case wide = "WIDE"

    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

File: shared/src/commonMain/kotlin/com/ansen/cms/shared/BlogPost.kt
package com.ansen.cms.shared

import kotlinx.serialization.Serializable

/**
 * The fields shared by the mobile clients and the server.
 * Dates use ISO-8601 calendar-date strings (for example, 2026-08-16).
 */
@Serializable
data class BlogPostDraft(
    val date: String,
    val title: String,
    val imageUrl: String? = null,
    val imageAltText: String? = null,
    val imageSize: ImageSize? = null,
    val subheading: String? = null,
    val link: String? = null,
    val linkText: String? = null,
    val bodyText: String,
)

@Serializable
data class BlogPost(
    val id: String,
    val date: String,
    val title: String,
    val imageUrl: String? = null,
    val imageAltText: String? = null,
    val imageSize: ImageSize? = null,
    val subheading: String? = null,
    val link: String? = null,
    val linkText: String? = null,
    val bodyText: String,
    val isPublished: Boolean,
)

@Serializable
enum class ImageSize {
    SMALL,
    MEDIUM,
    LARGE,
    WIDE,
}

fun BlogPostDraft.toPost(id: String, isPublished: Boolean) = BlogPost(
    id = id,
    date = date.trim(),
    title = title.trim(),
    imageUrl = imageUrl?.trim()?.ifBlank { null },
    imageAltText = imageAltText?.trim()?.ifBlank { null },
    imageSize = imageSize,
    subheading = subheading?.trim()?.ifBlank { null },
    link = link?.trim()?.ifBlank { null },
    linkText = linkText?.trim()?.ifBlank { null },
    bodyText = bodyText.trim(),
    isPublished = isPublished,
)
